### ISSUE TITLE
New types for IP addresses

### ISSUE THREAD

**Issue by dhess** [dhall-lang/dhall-lang#217]
Given:

- dhall is a configuration language
- IP addresses show up frequently in configuration
- There is already parser support for IP addresses

what do you think about making a Dhall-level type for IP addresses, both IPv4 and IPv6?

**Comment by Gabriel439** [comment]
Yeah, I think this is a good idea

I want to note that this is somewhat similar to another request for providing built-in support for dates and times: https://github.com/dhall-lang/dhall-lang/issues/80 but I think IP addresses are easier to add to the language because:

* The Dhall standard already includes the grammar rules for parsing them (like you mentioned)
    * I also don't think they conflict with any other elements of the grammar
* They have a smaller API surface than timestamps
    * Basically just a new `IP/show : IP → Text` built-in as far as I can tell

Did you want to contribute a pull request to propose the change?  Otherwise, I can write up the proposal if you want

**Comment by dhess** [comment]
I ran across #80 yesterday and thought you might object to this one on similar grounds, but I'm happy you agree!

Regarding the process, as this will be my first contribution to Dhall, how about you draft the proposal and I will work on the implementation?

**Comment by dhess** [comment]
To give you an idea of what I was thinking of, it would be useful to support both IPv4 and IPv6 addresses as separate types. Ideally they would also support CIDR notation, for representing subnets or IP ranges.

**Comment by dhess** [comment]
@Profpatsch Couldn't the same thing be said about HTTP URIs?

**Comment by Profpatsch** [comment]
They are resolved in the importer though and don’t have a representation in the dhall AST.

**Comment by Gabriel439** [comment]
Alright, then let's hold off on this

Another thing that just occurred to me is that if we provide built-in support for IP addresses then we need to pick a matching data type for IP addresses in each language that we bind Dhall to.  So I think that even if we were to do this we should probably wait until we have more than one language binding to see if that would be problematic.

**Comment by hanshoglund** [comment]
This seems like a special case of Vector and Word types.

Of course, if Vector existed, words of any size could be encoded as a Bool vector.
```haskell
let Vector : Natural -> Type -> Type
let Word8 = Vector 8 Bool
let IPv4 = Vector 4 Word8
let subnet = [255,255,255,0] : IPv4
```
And this can be faked now by encoding vectors as records
```haskell
let Vec2 = \(a : Type) -> {_1:a, _2:a }
let Vec3 = ...
```

**Comment by ocharles** [comment]
The ergonomics of all of those are still worse than a true IPv4 type. Dhall is meant to be a configuration language, and having to enter an IP as `{ _1 = 127, _2 = 0, _3 = 0, _4 = 1 }` is much worse than just `127.0.0.1`. In fact, so much worse that I think the best option right now `\(ip : Type) -> (mk-ip : String -> ip) -> ip` is a better type for IPs. (With `ip : String -> (ip : Type) -> (mk-ip : String -> ip) -> ip` so you can do `./ip "127.0.0.1"`.

**Comment by hanshoglund** [comment]
@ocharles Sure. I tend to prefer structural types to primitive ones, though I agree syntax is a problem :).

**Comment by hanshoglund** [comment]
I'd like to see Dhall develop in a direction where the type system stays mostly as it is but syntactic sugar is added for convenience. 

I wouldn't really mind a built in syntax for IPs/dates etc, as long as they resolve to structured types, e.g. making `127.0.0.1 : IPv4` valid syntax, but having `IPv4 = Vect 4 Word8` (or `IPv4 = Word32`). 

This would solve the "host language mapping" problem as well: structured types work in any language bindings, and more idiomatic specific host-language types can be added as requested. E.g. in Haskell you could unmarshall a dhall `IPv4` as `Data.Word.Word32`, `Network.Socket.HostAddr`, etc.

**Comment by duairc** [comment]
For what it's worth I made a "library" for IP addresses in Dhall [here](https://github.com/duairc/dhall-ip-address). They're correct by construction but the underlying implementation is pretty ugly.

The use case I originally had in mind was configuring a DNS server. I wanted to be able to automatically generate reverse records at the same time as A/AAAA records (in certain cases). That means I need more than just `IP{v4,v6}/show`, because I need to be able to pull them apart to get things like `4.4.8.8.in-addr.arpa` and `b.a.9.8.7.6.5.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.8.b.d.0.1.0.0.2.ip6.arpa`.

**Comment by andrewthad** [comment]
Dumping out some thoughts on this since I've been pondering today. One option would be to just treat an IP address as `Natural`. It wouldn't impact the type system at all, but it would have some kind of crummy properties. Adding IP addresses doesn't really make sense. And supporting masks like `192.0.2.0/24` would be useful, but desugaring to naturals does nothing to help with that. There's the suggestion made in https://github.com/dhall-lang/dhall-lang/issues/217#issuecomment-453884776:

> I wouldn't really mind a built in syntax for IPs/dates etc, as long as they resolve to structured types, e.g. making `127.0.0.1 : IPv4` valid syntax, but having `IPv4 = Vect 4 Word8 (or IPv4 = Word32)`.

The fixed-length vector seems reasonable, and subnets could be desugared pretty naturally as `{address : Vect 4 Word8, mask : Natural}` or something like that. But the whole `Word8` business seems a little nasty. I actually kind of like that dhall only supports unlimited-precision naturals. So perhaps `Vect 4 Natural` would be better. I don't really know.

Another alternative would be some kind of type-level tagging combined with the plain-old-Natural. Kind of like what the Haskell library `tagged` does. So something like `192.0.2.53 : Tagged "ipv4" Natural`. And then you would need functions `tag : forall (s :: String) -> forall (a :: Type) -> a -> Tag s a` and `untag : forall (s :: String) -> forall (a :: Type) -> Tag s a -> a`. The idea would be that users could just tag naturals that represent logically different things with different strings, and then untag them do actually do stuff with the underlying natural. This is kind of like Haskell's newtype mechanism, but it doesn't actually let people create new types.

Again, just a bunch of random thoughts from this morning.

**Comment by saurabhnanda** [comment]
Trying Dhall for the first time in a real-world configuration scenario (finally found one!), and I ran into this problem. What's the current recommendation for constructing (reasonably) safe/correct IP addresses in Dhall?

**Comment by Gabriel439** [comment]
@saurabhnanda: There isn't an alternative that is much better than `Text` at the moment

**Comment by ehmry** [comment]
FWIW people have been writting IPs in Nix for years and I doubt anyone has asked for a builtin. https://nixos.org/manual/nixos/stable/index.html#sec-ipv4
